actions/q-131: ADD Calling reusable workflow vs composite actions (#627)
FidelusAleksander/ghcertified: questions/en/actions/question-131.md

---
question: "Which of the following are true regarding calling reusable workflows versus calling composite actions?"
documentation: "https://docs.github.com/en/actions/concepts/workflows-and-actions/reusing-workflow-configurations#key-differences-between-reusable-workflows-and-composite-actions"
---

- [x] Composite actions are called via referencing the folder that contains their `action.yml` file.

> > As an action, composite actions must contain the brunt of their logic within an `action.yml` file. To call the composite action, point to where its `action.yml` is located (this includes the root. ex. to call a composite action that is located in the same repository as the caller workflow, the syntax `uses: ./` would be used).

Reviewer: ```suggestion
> As an action, composite actions must contain the brunt of their logic within an `action.yml` file. To call the composite action, point to where its `action.yml` is located (this includes the root. ex. to call a composite action that is located at the root of the same repository as the caller workflow, the syntax `uses: ./` would be used).
```

- [ ] Reusable workflows are called via referencing the folder that contains their `action.yml` file.
> Reusable workflows are regular `.yml` or `.yaml` files that are stored in `.github/workflows`. They do not have an `action.yml` file.

> - [ ] Composite actions must be called as a step within a job (not from step-level).

Reviewer: This is true now after the changes
```suggestion
- [x] Composite actions must be called as a step within a job
```

> > Composite actions (as with any other action) are called from within a __step__ of a workflow job--in other words, you do not need a specific workflow job just to caller a composite action. 

Reviewer: ```suggestion
> Composite actions (as with any other action) are called from within a step of a workflow job--in other words, you do not need a specific workflow job just to caller a composite action. 
```

> - [x] Reusable workflows must be called directly within a workflow job (not from step-level).

Reviewer: for clarity since steps are also "within a workflow job" 
```suggestion
- [x] Reusable workflows must be called on workflow job level (not from step-level).
```

> > Steps within a workflow job cannot call a reusable workflow. A reusable workflow must be called by an indvidiual job within the caller workflow. This can result in one or more jobs running in the caller workflow (said jobs can be seen in workflow runs in the Github Actions UI). 

Reviewer: ```suggestion
> Steps within a workflow job cannot call a reusable workflow. A reusable workflow must be called by an individual job within the caller workflow. This can result in one or more jobs running in the caller workflow (said jobs can be seen in workflow runs in the Github Actions UI). 
```

> 

Reviewer: Remove this extra whitespace

Author: Commited

---
question: "Which of the following are true regarding calling reusable workflows versus calling composite actions?"
documentation: "https://docs.github.com/en/actions/concepts/workflows-and-actions/reusing-workflow-configurations#key-differences-between-reusable-workflows-and-composite-actions"
---

- [x] Composite actions are called via referencing the folder that contains their `action.yml` file.

> > As an action, composite actions must contain the brunt of their logic within an `action.yml` file. The composite action's `action.yml` must reside in an `.github/actions/<action-name>` subfolder.

Reviewer: > The composite action's `action.yml` must reside in an `.github/actions/<action-name>` subfolder.

This is not true. It can be any folder. Also multiple repositories contain composite actions at the root directory so you can call them exactly as other actions. 

Composite actions are called exactly the same way as any other actions

Author: Shoot, this I didn't know, I thought it had to be the specific folder '.github/actions'. Thanks!

Author: Committed. Let me know if the new explanation sounds good

- [ ] Reusable workflows are called via referencing the folder that contains their `action.yml` file.
> Reusable workflows are regular `.yml` or `.yaml` files that are stored in `.github/workflows`. They do not have an `action.yml` file.

> - [ ] Composite actions must be called directly from a job.

Reviewer: Honestly the way this is written it does kinda sound like it's true. 

Actions are called FROM a job - but AS A step

```suggestion
- [ ] Composite actions must be called as a step
```

Author: Committed

Ugh yeah I hate the way the docs word it. I used the section "Key differences between reusable workflows and composite actions" section on [this page](https://docs.github.com/en/actions/concepts/workflows-and-actions/reusing-workflow-configurations#key-differences-between-reusable-workflows-and-composite-actions) and tried to not directly take its phrasing, which didn't work haha. I've amended it to be more explicit.

Reviewer: @mcummings128 I don't think you've actually pushed these changes?

Author: Thanks, I didn't realize! I just pushed them now

> Composite actions (as with any other action) are called from within a __step__ of a workflow job--in other words, you do not need a specific workflow job just to caller a composite action. 

> - [x] Reusable workflows must be called directly from a job.

Reviewer: Same comment as above. Reusable workflows are called from different workflows, not from jobs. 

Also when you fix that a reusable workflow call can result in one or more jobs running in the caller workflow

Author: Committed

Similar to the above, I used the section "Key differences between reusable workflows and composite actions" section on [this page](https://docs.github.com/en/actions/concepts/workflows-and-actions/reusing-workflow-configurations#key-differences-between-reusable-workflows-and-composite-actions). That phrasing for reusable workflows says "directly within a job" but I've modified these two answers to be more explicit.

Also added the note about one or more jobs running in the caller workflow, additionally mentioning that these can be seen explicitly in the Github Actions UI.

Let me know if this still needs edits. Thanks!

Reviewer: @mcummings128 I don't think you've actually pushed these changes?

Author: Thanks, I didn't realize! I just pushed them now
